docs/types.rst

```diff
@@ -413,8 +413,18 @@ function type should not return anything, the whole ``returns (<return types>)``
 part has to be omitted.
 
 By default, function types are internal, so the ``internal`` keyword can be
-omitted. In contrast, contract functions themselves are public by default,
-only when used as the name of a type, the default is internal.
+omitted. In contrast, visibility has to be specified for contract functions themselves,
+only when used as a type, there is a default of internal.
+
+A function type ``A`` is implicitly convertible to a function type ``B`` if and only if
+their parameter types are identical, their return types are identical,
+their internal/external property is identical and the state mutability of ``A``
+is not more restrictive than the state mutability of ``B``. In this sense,
+``view`` functions are considered most restrictive and ``payable`` least
```

Comment:
This should probably be ``pure`` not ``view``.
Maybe we should just mention the order ``pure > view > payable > nonpayable``.

Reply:
Hm - the order is actually ``pure > view > nonpayable`` and ``payable > nonpayable``, but not ``view > payable``.

Reply:
Right, non-payable is least restrictive.

Reply:
Maybe we should divide this into ``and the state mutability of A is not more restrictive than the state mutability of B, except if one of them is payable. A payable function type can be converted to a nonpayable function type, but no type that is not also payable can be converted to a payable function type.``

...or something better - in any case I would make the entire order explicit here.

Reply:
Maybe we should even put ``you are free to pay zero ether`` and ``sending ether to non-payable is not allowed`` in parenthesis here.